spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/GqlDatastoreQuery.java

```diff
@@ -112,26 +115,78 @@ public GqlDatastoreQuery(Class<T> type, DatastoreQueryMethod queryMethod,
 	@Override
 	public Object execute(Object[] parameters) {
 		GqlQuery query = bindArgsToGqlQuery(this.gql, getParamTags(), parameters);
-		Iterable found = this.datastoreTemplate.queryKeysOrEntities(query, this.entityType);
+		boolean returnTypeIsCollection = this.queryMethod.isCollectionQuery();
+		Class returnedItemType = returnTypeIsCollection
+				? this.queryMethod.getResultProcessor().getReturnedType()
+						.getReturnedType()
+				: this.queryMethod.getReturnedObjectType();
+		boolean returnItemTypeIsNonEntity = isNonEntityReturnedType(returnedItemType);
+
+		Iterable found = returnItemTypeIsNonEntity
+				? this.datastoreTemplate.query(query, x -> {
+					Object mappedResult;
+					if (x instanceof Key) {
+						mappedResult = x;
+					}
+					else {
+						BaseEntity entity = (BaseEntity) x;
+						mappedResult = entity
+								.getValue((String) entity.getNames().toArray()[0]).get();
+					}
+					return mappedResult;
+				})
+				: this.datastoreTemplate.queryKeysOrEntities(query, this.entityType);
+
 		List rawResult = found == null ? Collections.emptyList()
 				: (List) StreamSupport.stream(found.spliterator(), false)
 						.collect(Collectors.toList());
 		Object result;
+
 		if (this.queryMethod.isCountQuery()) {
 			result = rawResult.size();
 		}
 		else if (this.queryMethod.isExistsQuery()) {
 			result = !rawResult.isEmpty();
 		}
-		else if (rawResult.isEmpty() || rawResult.get(0).getClass() == Key.class) {
-			result = rawResult;
+		else if (returnItemTypeIsNonEntity) {
+			if (returnTypeIsCollection) {
+				result = rawResult.stream()
+						.map(x -> this.datastoreTemplate.getDatastoreEntityConverter()
+								.getConversions().getConversionService()
+								.convert(x, returnedItemType))
+						.collect(Collectors.toList());
```

Comment:
I don't think we should assume that collections are always Lists; queryMethod.isCollectionQuery() returns true for arrays, specific List implementations, Sets, etc

If possible, it would make sense to use conversionService to convert results list into required collection type.